flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableInputFormat.java

```diff
@@ -0,0 +1,327 @@
+			SerDeUtils.initializeSerDe(deserializer, conf, properties, null);
+			// Get the row structure
+			oi = (StructObjectInspector) deserializer.getObjectInspector();
+			fieldRefs = oi.getAllStructFieldRefs();
+		} catch (Exception e) {
+			logger.error("Error happens when deserialize from storage file.");
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Override
+	public HiveTableInputSplit[] createInputSplits(int minNumSplits)
+			throws IOException {
+		List<HiveTableInputSplit> hiSplit = new ArrayList<>();
+		int splitNum = 0;
+		for (HiveTablePartition partition : partitions) {
+			StorageDescriptor sd = partition.getStorageDescriptor();
+			InputFormat format;
+			try {
+				format = (InputFormat)
+					Class.forName(sd.getInputFormat(), true, Thread.currentThread().getContextClassLoader()).newInstance();
+			} catch (Exception e) {
+				throw new RuntimeException("Unable to instantiate the hadoop input format", e);
+			}
+			ReflectionUtils.setConf(format, jobConf);
+			jobConf.set(INPUT_DIR, sd.getLocation());
+			//TODO: we should consider how to calculate the splits according to minNumSplits in the future.
+			org.apache.hadoop.mapred.InputSplit[] splitArray = format.getSplits(jobConf, minNumSplits);
+			for (int i = 0; i < splitArray.length; i++) {
+				hiSplit.add(new HiveTableInputSplit(splitNum++, splitArray[i], jobConf, partition));
+			}
+		}
+
+		return hiSplit.toArray(new HiveTableInputSplit[hiSplit.size()]);
+	}
+
+	@Override
+	public void configure(org.apache.flink.configuration.Configuration parameters) {
+
+	}
+
+	// Todo: refactor code to get statistics from hms or partition.
+	@Override
+	public BaseStatistics getStatistics(BaseStatistics cachedStats) throws IOException {
+		// only gather base statistics for FileInputFormats
+		if (!(mapredInputFormat instanceof FileInputFormat)) {
+			return null;
+		}
+
+		final org.apache.flink.api.common.io.FileInputFormat.FileBaseStatistics cachedFileStats =
+				(cachedStats instanceof org.apache.flink.api.common.io.FileInputFormat.FileBaseStatistics) ?
+				(org.apache.flink.api.common.io.FileInputFormat.FileBaseStatistics) cachedStats : null;
+
+		try {
+			final org.apache.hadoop.fs.Path[] paths = FileInputFormat.getInputPaths(this.jobConf);
+
+			return HiveTableUtil.getFileStats(cachedFileStats, paths, new ArrayList<FileStatus>(1));
+		} catch (IOException ioex) {
+			if (logger.isWarnEnabled()) {
+				logger.warn("Could not determine statistics due to an io error: "
+							+ ioex.getMessage());
+			}
+		} catch (Throwable t) {
+			if (logger.isErrorEnabled()) {
+				logger.error("Unexpected problem while getting the file statistics: " + t.getMessage(), t);
+			}
+		}
+
+		// no statistics available
+		return null;
+	}
+
+	@Override
+	public InputSplitAssigner getInputSplitAssigner(HiveTableInputSplit[] inputSplits) {
+		return new LocatableInputSplitAssigner(inputSplits);
+	}
+
+	@Override
+	public boolean reachedEnd() throws IOException {
+		if (!fetched) {
```

Comment:
is `fetched` dup of `hasNext`?

Reply:
nope. reachedEnd() method is called as a loop condition while  nextRecord() is the loop body.
`fetched` will change after nextRecord() called, while `hasNext` only change after `fetchNext()` called.

Reply:
I'm still a bit confused by the logic of fetching and hasNext.

As as you said `reachEnd()` is looping condition and `nextRecord()` is loop body, then we don't even need to worry about `hasNext` in `nextRecord()` anymore, right? And it can simplified as:


```
@Override
public boolean reachedEnd() throws IOException {
	return !this.recordReader.next(key, value);
}

@Override
public Row nextRecord(Row ignore) throws IOException {
	Row row = new Row(rowTypeInfo.getArity());
	.... 
}
```

Reply:
it's the inner implementation that taking `reachEnd()` as looping condition. While the interface of `InputFormat`  doesn't guarantee the calling method. that's why we need a `fetchNext` method. Which is also a reference from `HadoopInputFormat`.

Reply:
ok, I read the `HCatInputFormatBase` and `HadoopInputFormatBase`, and yes, they are implemented this way.